File: plugin/src/knowledge/providers/sourceforge.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const SOURCEFORGE_API = "https://sourceforge.net/api";

interface SourceForgeSearchResult {
  projects?: Array<{
    projectid: number;
    name: string;
    url: string;
    shortname: string;
    summary: string;
    categories: {
      topic: Array<{ fullname: string }>;
      os: Array<{ fullname: string }>;
      language: Array<{ fullname: string }>;
      license: Array<{ fullname: string }>;
    };
  }>;
}

interface SourceForgeProject {
  name: string;
  shortname: string;
  summary: string;
  description?: string;
  url: string;
  created: string;
  homepage?: string;
  external_homepage?: string;
  download_page?: string;
  categories: {
    topic?: Array<{ fullname: string }>;
    os?: Array<{ fullname: string }>;
    language?: Array<{ fullname: string }>;
    license?: Array<{ fullname: string }>;
  };
  developers?: Array<{
    username: string;
    name: string;
  }>;
  stats?: {
    downloads?: number;
  };
}

export class SourceForgeProvider implements KnowledgeProvider {
  readonly name = "sourceforge";

  async isAvailable(): Promise<boolean> {
    try {
      const response = await requestUrl({
        url: `${SOURCEFORGE_API}/project/name/test/json`,
        method: "GET",
        throw: false,
      });
      // 404 is ok, means API is responding
      return response.status === 200 || response.status === 404;
    } catch {
      return false;
    }
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    try {
      // Check if query looks like a project name
      if (!query.includes(" ") && query.length < 50) {
        const project = await this.lookupProject(query);
        if (project) {
          return {
            success: true,
            provider: this.name,
            entries: [project],
          };
        }
      }

      // Otherwise search
      const entries = await this.search(query, maxResults);

      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `SourceForge lookup failed: ${message}`,
      };
    }
  }

  private async lookupProject(name: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${SOURCEFORGE_API}/project/name/${encodeURIComponent(name)}/json`,
      method: "GET",
      throw: false,
    });

    if (response.status === 404) {
      return null;
    }

    if (response.status !== 200) {
      throw new Error(`Project lookup failed: ${response.status}`);
    }

    const data = response.json as { Project?: SourceForgeProject };
    const project = data.Project;

    if (!project) {
      return null;
    }

    return this.projectToEntry(project);
  }

  private async search(query: string, limit: number): Promise<KnowledgeEntry[]> {
    // SourceForge doesn't have a great search API, use their sitemap/allura search
    const response = await requestUrl({
      url: `https://sourceforge.net/directory/?q=${encodeURIComponent(query)}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      // Fallback: try direct project lookup
      const project = await this.lookupProject(query.replace(/\s+/g, "").toLowerCase());
      return project ? [project] : [];
    }

    // Parse HTML for project links (basic extraction)
    const projects = this.extractProjectsFromHtml(response.text, limit);

    // Fetch details for each project
    const entries: KnowledgeEntry[] = [];
    for (const shortname of projects) {
      const entry = await this.lookupProject(shortname);
      if (entry) {
        entries.push(entry);
        if (entries.length >= limit) break;
      }
    }

    return entries;
  }

  private extractProjectsFromHtml(html: string, limit: number): string[] {
    const projects: string[] = [];
    const regex = /href="\/projects\/([a-zA-Z0-9_-]+)\/"/g;
    let match;

    while ((match = regex.exec(html)) !== null && projects.length < limit * 2) {
      const shortname = match[1];
      if (!projects.includes(shortname) && shortname !== "directory") {
        projects.push(shortname);
      }
    }

    return projects.slice(0, limit);
  }

  private projectToEntry(project: SourceForgeProject): KnowledgeEntry {
    const lines: string[] = [];

    if (project.summary) {
      lines.push(project.summary);
    }

    const languages = project.categories.language?.map((l) => l.fullname).slice(0, 3);
    if (languages && languages.length > 0) {
      lines.push(`Languages: ${languages.join(", ")}`);
    }

    const licenses = project.categories.license?.map((l) => l.fullname).slice(0, 2);
    if (licenses && licenses.length > 0) {
      lines.push(`License: ${licenses.join(", ")}`);
    }

    const os = project.categories.os?.map((o) => o.fullname).slice(0, 3);
    if (os && os.length > 0) {
      lines.push(`Platforms: ${os.join(", ")}`);
    }

    return {
      title: project.name,
      summary: lines.join("\n"),
      url: project.url || `https://sourceforge.net/projects/${project.shortname}/`,
      source: "sourceforge",
      metadata: {
        shortname: project.shortname,
        description: project.description,
        homepage: project.external_homepage || project.homepage,
        downloadPage: project.download_page,
        topics: project.categories.topic?.map((t) => t.fullname),
        languages: project.categories.language?.map((l) => l.fullname),
        licenses: project.categories.license?.map((l) => l.fullname),
        platforms: project.categories.os?.map((o) => o.fullname),
        developers: project.developers?.map((d) => d.name || d.username),
        downloads: project.stats?.downloads,
        created: project.created,
      },
    };
  }
}

File: crates/mcp/src/knowledge/sourceforge.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const SOURCEFORGE_API: &str = "https://sourceforge.net/api";

pub struct SourceForgeProvider {
    client: Client,
}

#[derive(Debug, Deserialize)]
struct ProjectResponse {
    #[serde(rename = "Project")]
    project: Option<Project>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct Project {
    name: String,
    shortname: String,
    summary: Option<String>,
    description: Option<String>,
    url: Option<String>,
    created: Option<String>,
    homepage: Option<String>,
    external_homepage: Option<String>,
    download_page: Option<String>,
    categories: Option<Categories>,
}

#[derive(Debug, Deserialize)]
struct Categories {
    topic: Option<Vec<Category>>,
    os: Option<Vec<Category>>,
    language: Option<Vec<Category>>,
    license: Option<Vec<Category>>,
}

#[derive(Debug, Deserialize)]
struct Category {
    fullname: String,
}

impl SourceForgeProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
        }
    }

    fn project_to_entry(&self, project: &Project) -> KnowledgeEntry {
        let mut lines = Vec::new();

        if let Some(summary) = &project.summary {
            lines.push(summary.clone());
        }

        if let Some(cats) = &project.categories {
            if let Some(langs) = &cats.language {
                let lang_names: Vec<_> = langs.iter().take(3).map(|c| c.fullname.clone()).collect();
                if !lang_names.is_empty() {
                    lines.push(format!("Languages: {}", lang_names.join(", ")));
                }
            }

            if let Some(licenses) = &cats.license {
                let lic_names: Vec<_> = licenses.iter().take(2).map(|c| c.fullname.clone()).collect();
                if !lic_names.is_empty() {
                    lines.push(format!("License: {}", lic_names.join(", ")));
                }
            }

            if let Some(os) = &cats.os {
                let os_names: Vec<_> = os.iter().take(3).map(|c| c.fullname.clone()).collect();
                if !os_names.is_empty() {
                    lines.push(format!("Platforms: {}", os_names.join(", ")));
                }
            }
        }

        let mut metadata = HashMap::new();
        metadata.insert("shortname".to_string(), serde_json::json!(project.shortname));
        if let Some(desc) = &project.description {
            metadata.insert("description".to_string(), serde_json::json!(desc));
        }
        if let Some(hp) = project.external_homepage.as_ref().or(project.homepage.as_ref()) {
            metadata.insert("homepage".to_string(), serde_json::json!(hp));
        }
        if let Some(cats) = &project.categories {
            if let Some(topics) = &cats.topic {
                let names: Vec<_> = topics.iter().map(|t| t.fullname.clone()).collect();
                metadata.insert("topics".to_string(), serde_json::json!(names));
            }
            if let Some(langs) = &cats.language {
                let names: Vec<_> = langs.iter().map(|l| l.fullname.clone()).collect();
                metadata.insert("languages".to_string(), serde_json::json!(names));
            }
        }

        let url = project.url.clone()
            .unwrap_or_else(|| format!("https://sourceforge.net/projects/{}/", project.shortname));

        KnowledgeEntry {
            title: project.name.clone(),
            summary: lines.join("\n"),
            url: Some(url),
            source: "sourceforge".to_string(),
            metadata: Some(metadata),
        }
    }

    fn lookup_project(&self, name: &str) -> Result<Option<KnowledgeEntry>, String> {
        let url = format!("{}/project/name/{}/json", SOURCEFORGE_API, urlencoding::encode(name));

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if response.status().as_u16() == 404 {
            return Ok(None);
        }

        if !response.status().is_success() {
            return Err(format!("project lookup failed: {}", response.status()));
        }

        let data: ProjectResponse = response.json().map_err(|e| e.to_string())?;

        Ok(data.project.as_ref().map(|p| self.project_to_entry(p)))
    }
}

impl Default for SourceForgeProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for SourceForgeProvider {
    fn name(&self) -> &'static str {
        "sourceforge"
    }

    fn is_available(&self) -> bool {
        self.client
            .get(&format!("{}/project/name/test/json", SOURCEFORGE_API))
            .send()
            .map(|r| r.status().is_success() || r.status().as_u16() == 404)
            .unwrap_or(false)
    }

    fn lookup(&self, query: &str, _options: &LookupOptions) -> LookupResult {
        // Try direct project lookup
        match self.lookup_project(query) {
            Ok(Some(entry)) => LookupResult::success(self.name(), vec![entry]),
            Ok(None) => {
                // Try with spaces replaced
                let normalized = query.replace(' ', "").to_lowercase();
                match self.lookup_project(&normalized) {
                    Ok(Some(entry)) => LookupResult::success(self.name(), vec![entry]),
                    Ok(None) => LookupResult::success(self.name(), vec![]),
                    Err(e) => LookupResult::error(self.name(), e),
                }
            }
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore] // Requires network
    fn sourceforge_lookup() {
        let provider = SourceForgeProvider::new();
        let result = provider.lookup("audacity", &LookupOptions::default());
        assert!(result.success);
    }
}
